ledger/participant-state/kvutils/src/main/scala/kvutils/com/daml/ledger/participant-state/kvutils/InMemoryKVParticipantState.scala

```diff
@@ -0,0 +1,143 @@
+    ByteString.copyFromUtf8(idx.toString)
+
+  private def kvEntryIdToIdx(entryId: KeyValueUtils.KVEntryId): Long =
+    entryId.toStringUtf8.toLong
+
+  override def stateUpdates(beginAfter: Option[Offset]): Source[(Offset, Update), NotUsed] = {
+    dispatcher
+      .startingAt(beginAfter.map(_.components.head.toLong).getOrElse(beginning))
+      .map {
+        case (idx, update) =>
+          Offset(Array(idx)) -> update
+      }
+  }
+
+  override def submitTransaction(
+      submitterInfo: SubmitterInfo,
+      transactionMeta: TransactionMeta,
+      transaction: SubmittedTransaction): Unit = {
+    val entry = KeyValueUtils.transactionToEntry(submitterInfo, transactionMeta, transaction)
+
+    // Simulate a validation of the transaction. If the validation
+    // fails we inject a rejected command, otherwise we store the blob.
+    val txInfo = KeyValueUtils.computeTxInfo(entry)
+    val stateSnapshot = stateRef.get
+
+    // FIXME(JM): Add some error handling
+    val inputEntries = txInfo.inputs.contracts
+      .map(_._1)
+      .map { entryId =>
+        entryId -> stateSnapshot.entries(kvEntryIdToIdx(entryId))._2
+      }
+      .toMap
+
+    KeyValueUtils.validateTransaction(
+      engine,
+      stateSnapshot.archives, // FIXME(JM): "txInfo.inputs.packages" aren't correctly computed yet.
+      /*
+      txInfo.inputs.packages.map { k =>
+        k -> stateSnapshot.archives(k)
+      }.toMap,*/
+      inputEntries,
+      txInfo
+    ) match {
+      case Left(err) =>
+        val rejection =
+          Update.CommandRejected(Some(submitterInfo), RejectionReason.Disputed(err.toString))
+        val updatedState = stateRef.updateAndGet { state =>
+          state.copy(
+            nextIdx = state.nextIdx + 1,
+            rejections = state.rejections + (state.nextIdx -> rejection)
+          )
+        }
+        dispatcher.signalNewHead(updatedState.nextIdx)
+
+      case Right(()) =>
+        val updatedState = stateRef.updateAndGet { state =>
+          state.copy(
+            nextIdx = state.nextIdx + 1,
+            entries = state.entries + (state.nextIdx -> ((state.recordTime, entry)))
+          )
+        }
+        dispatcher.signalNewHead(updatedState.nextIdx)
+    }
+  }
+
+  // Back-channel for uploading DAML-LF archives.
+  def uploadArchive(archive: Archive): Unit = {
+    val entry = KeyValueUtils.archiveToEntry(archive)
+    val updatedState = stateRef.updateAndGet { state =>
+      state.copy(
+        nextIdx = state.nextIdx + 1,
+        entries = state.entries + (state.nextIdx -> ((state.recordTime, entry))),
+        archives = state.archives + (SimpleString
+          .assertFromString(archive.getHash) -> archive.toByteString)
+      )
+    }
+    dispatcher.signalNewHead(updatedState.nextIdx)
+  }
+
+  /** Retrieve the static initial conditions of the ledger, containing
+    * the ledger identifier and the epoch of the ledger record time.
```

Comment:
why "epoch of the ledger record time"; wouldn't just "ledger record time" be as precise?

Reply:
and less confusing... to me at least